scalameta/trees/shared/src/main/scala/scala/meta/internal/trees/InternalTrees.scala

```diff
@@ -64,19 +64,24 @@ trait InternalTree extends Product {
   }
 
   def tokenizeFor(dialect: Dialect): Tokens =
-    if (origin.dialectOpt.contains(dialect)) tokensOpt.get else retokenize(dialect, None)
+    if (origin.dialectOpt.contains(dialect)) tokensOpt.get else retokenizeFor(dialect)
 
   private def tokensOpt: Option[Tokens] = origin.tokensOpt.orElse(syntaxTokensOpt)
-  private lazy val syntaxTokensOpt: Option[Tokens] = origin.dialectOpt.map(retokenize(_, None))
+  private lazy val syntaxTokensOpt: Option[Tokens] = origin.dialectOpt.map(retokenizeFor)
 
   private[meta] def retokenize(implicit
+      tokenize: Tokenize,
       dialect: Dialect,
       tokenizerOptions: Option[TokenizerOptions]
-  ): Tokens = this match {
+  ): Tokens = retokenizeFor(dialect)
+
+  private[meta] def retokenizeFor(
+      dialect: Dialect
+  )(implicit tokenize: Tokenize, tokenizerOptions: Option[TokenizerOptions]): Tokens = this match {
     case Lit.String(value) =>
       val input = Input.VirtualFile("<InternalTrees.tokens>", value)
       Tokens(Array(Constant.String(input, dialect, 0, value.length, value)))
-    case _ => implicitly[Tokenize].apply(textAsInput, dialect).get
+    case _ => tokenize(origin.inputOpt.getOrElse(textAsInput), dialect).get
   }
 
   private[meta] def textAsInput(implicit
```

Comment:
not visible in the diff, but another implicit `Option[TokenizerOptions]` is captured - it's probably more correct to get it from the input just like above?